presto-main/src/main/java/com/facebook/presto/operator/MergeOperator.java

```diff
@@ -0,0 +1,338 @@
+    {
+        private final int operatorId;
+        private final PlanNodeId sourceId;
+        private final ExchangeClientSupplier exchangeClientSupplier;
+        private final PagesSerdeFactory serdeFactory;
+        private final List<Type> sourceTypes;
+        private final List<Integer> outputChannels;
+        private final List<Type> outputTypes;
+        private final List<Integer> sortChannels;
+        private final List<SortOrder> sortOrder;
+        private final MergeSortComparatorFactory mergeSortComparatorFactory;
+        private boolean closed;
+
+        public MergeOperatorFactory(
+                int operatorId,
+                PlanNodeId sourceId,
+                ExchangeClientSupplier exchangeClientSupplier,
+                PagesSerdeFactory serdeFactory,
+                MergeSortComparatorFactory mergeSortComparatorFactory,
+                List<Type> sourceTypes,
+                List<Integer> outputChannels,
+                List<Integer> sortChannels,
+                List<SortOrder> sortOrder)
+        {
+            this.operatorId = operatorId;
+            this.sourceId = requireNonNull(sourceId, "sourceId is null");
+            this.exchangeClientSupplier = requireNonNull(exchangeClientSupplier, "exchangeClientSupplier is null");
+            this.serdeFactory = requireNonNull(serdeFactory, "serdeFactory is null");
+            this.sourceTypes = requireNonNull(sourceTypes, "sourceTypes is null");
+            this.outputChannels = requireNonNull(outputChannels, "outputChannels is null");
+            this.outputTypes = outputTypes(sourceTypes, outputChannels);
+            this.sortChannels = requireNonNull(sortChannels, "sortChannels is null");
+            this.sortOrder = requireNonNull(sortOrder, "sortOrder is null");
+            this.mergeSortComparatorFactory = requireNonNull(mergeSortComparatorFactory, "mergeSortComparatorFactory is null");
+        }
+
+        private static List<Type> outputTypes(List<? extends Type> sourceTypes, List<Integer> outputChannels)
+        {
+            ImmutableList.Builder<Type> types = ImmutableList.builder();
+            for (int channel : outputChannels) {
+                types.add(sourceTypes.get(channel));
+            }
+            return types.build();
+        }
+
+        private static List<Type> sortTypes(List<? extends Type> sourceTypes, List<Integer> sortChannels)
+        {
+            ImmutableList.Builder<Type> types = ImmutableList.builder();
+            for (int channel : sortChannels) {
+                types.add(sourceTypes.get(channel));
+            }
+            return types.build();
+        }
+
+        @Override
+        public PlanNodeId getSourceId()
+        {
+            return sourceId;
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return outputTypes;
+        }
+
+        @Override
+        public SourceOperator createOperator(DriverContext driverContext)
```

Comment:
The class is being considered non thread safe if it is not marked as `ThreadSafe` explicitly. See `HashBuilderOperatorFactory`.